Process query from liaison

lujiajing1126 opened the pull request:
A `Write` API is added to support writing entities when running tests.

- [x] TraceIDFetch Test
- [x] TableScan Test

Fine0830 commented:
Please give the junior player more time. 😊

lujiajing1126 commented:
> Please give the junior player more time. 😊

No rush, pls take ur time! Piano Piano

lujiajing1126 commented:
@hanahmily @Fine0830 Pls have a review.

But the test seems not stable, though `All checks have passed`. I am still investigating the reason, any idea @hanahmily ?

```
2021-07-30T09:29:13+08:00 | ERROR | ***Failure while flushing memtable to disk: error while creating table err: while creating table: /tmp/sw/shard-0/seg-20210730/block-0929/chunkIDMapping/000001.sst error: open /tmp/sw/shard-0/seg-20210730/block-0929/chunkIDMapping/000001.sst: no such file or directory
unable to open: /tmp/sw/shard-0/seg-20210730/block-0929/chunkIDMapping/000001.sst
github.com/dgraph-io/ristretto/z.OpenMmapFile
	/Users/megrez/go/pkg/mod/github.com/dgraph-io/ristretto@v0.0.4-0.20210504190834-0bf2acd73aa3/z/file.go:80
github.com/dgraph-io/badger/v3/table.newFile
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/table/table.go:274
github.com/dgraph-io/badger/v3/table.CreateTable
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/table/table.go:260
github.com/dgraph-io/badger/v3.(*DB).handleFlushTask
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:1127
github.com/dgraph-io/badger/v3.(*DB).flushMemtable
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:1190
github.com/dgraph-io/badger/v3.Open.func5
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:362
runtime.goexit
	/opt/homebrew/Cellar/go/1.16.6/libexec/src/runtime/asm_arm64.s:1130. Retrying...
```

and

```
2021-07-30T09:29:15+08:00 | ERROR | ***while deleting file: /tmp/sw/shard-0/seg-20210730/block-0929/chunkIDMapping/00001.mem, err: while munmap file: /tmp/sw/shard-0/seg-20210730/block-0929/chunkIDMapping/00001.mem, error: invalid argument
**** module:BLOCK.NORMAL-KV
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x10 pc=0x1026dfe64]

goroutine 12 [running]:
github.com/dgraph-io/badger/v3/skl.(*Arena).getNode(...)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/skl/arena.go:126
github.com/dgraph-io/badger/v3/skl.(*Skiplist).getHead(...)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/skl/skl.go:191
github.com/dgraph-io/badger/v3/skl.(*Iterator).SeekToFirst(...)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/skl/skl.go:481
github.com/dgraph-io/badger/v3/skl.(*UniIterator).Rewind(0x1400030b5a0)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/skl/skl.go:518 +0xa4
github.com/dgraph-io/badger/v3.buildL0Table(0x0, 0x0, 0x1029fdba8, 0x1400030b5a0, 0x0, 0x0, 0x0, 0x100, 0x200000, 0x0, ...)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:1084 +0xa0
github.com/dgraph-io/badger/v3.(*DB).handleFlushTask(0x14000262480, 0x0, 0x0, 0x1029fdba8, 0x1400030b5a0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:1109 +0xe0
github.com/dgraph-io/badger/v3.(*DB).flushMemtable(0x14000262480, 0x140006780f0, 0x0, 0x0)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:1190 +0x328
github.com/dgraph-io/badger/v3.Open.func5(0x140001304c0)
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:362 +0x34
created by github.com/dgraph-io/badger/v3.Open
	/Users/megrez/go/pkg/mod/github.com/!sky!a!p!m/badger/v3@v3.0.0-20210624023741-bd2dcfcaaa74/db.go:361 +0x928
FAIL	github.com/apache/skywalking-banyandb/banyand/query	4.111s
```

hanahmily commented:
> @hanahmily @Fine0830 Pls have a review.
> 
> But the test seems not stable, though `All checks have passed`. I am still investigating the reason, any idea @hanahmily ?
> 

That's due to the incomplete storage module which can't handle existing data files for now. My suggestion is to remove the entire data directory, `/tmp/sw` in your case, after the test. You can stop database object gracefully by referring to https://github.com/apache/skywalking-banyandb/blob/main/banyand/series/trace/query_test.go#L140

hanahmily commented on banyand/query/processor.go:
```diff
@@ -0,0 +1,96 @@
+		Spec:        queryCriteria.GetMetadata(),
+	}
+	s, err := analyzer.BuildTraceSchema(context.TODO(), *metadata)
+	if err != nil {
```
Do we have to log this error if we don't handle it under the hood?

lujiajing1126 replied:
Fixed

lujiajing1126 replied:
I suppose later we may improve bus component to add support for error handling

hanahmily commented on pkg/query/logical/analyzer.go:
```diff
@@ -94,14 +95,17 @@ func (a *Analyzer) Analyze(_ context.Context, criteria *apiv1.QueryRequest, trac
 			typedPair := pairQuery.GetCondition()
 			switch v := typedPair.GetTyped().(type) {
 			case *apiv1.TypedPair_StrPair:
+				// TODO: use metadata from Plugin?
```
Retrieving the key from https://github.com/apache/skywalking-banyandb/blob/main/banyand/series/schema/sw/trace_series.textproto#L75

lujiajing1126 replied:
Fixed. Extra functions are added to the Schema.

hanahmily commented on pkg/query/logical/plan_indexscan.go:
```diff
@@ -107,31 +107,49 @@ type indexScan struct {
+				chunkSet = chunks
+			} else {
+				// afterwards, it must not be nil
+				_ = chunkSet.Intersect(chunks)
```
Once the result of the intersection is empty, you can stop the loop immediately.

lujiajing1126 replied:
Fixed

hanahmily commented on pkg/query/logical/plan_indexscan.go:
```diff
@@ -107,31 +107,49 @@ type indexScan struct {
+	dataEntities := make([]data.Entity, 0)
+
+	// iterate over shards
+	for shardID := uint32(0); shardID < i.schema.ShardNumber(); shardID++ {
```
Can we stop scanning when the limitation of results is reached?

lujiajing1126 replied:
> Can we stop scanning when the limitation of results is reached?

I don't think so. If we want to do that,  we have to push down `Limit` as well as `Offset`.

hanahmily replied:
It's a performance issue more than a critical bug, we can improve it later. But we should mark it in the codes(TODO or something else) in case of missing it.

lujiajing1126 replied:
> It's a performance issue more than a critical bug, we can improve it later. But we should mark it in the codes(TODO or something else) in case of missing it.

Sure. I've added a TODO there

hanahmily reviewed (commented):
Looks great, some improvement should be taken into consideration though.

lujiajing1126 commented:
> > @hanahmily @Fine0830 Pls have a review.
> > But the test seems not stable, though `All checks have passed`. I am still investigating the reason, any idea @hanahmily ?
> 
> That's due to the incomplete storage module which can't handle existing data files for now. My suggestion is to remove the entire data directory, `/tmp/sw` in your case, after the test. You can stop database object gracefully by referring to https://github.com/apache/skywalking-banyandb/blob/main/banyand/series/trace/query_test.go#L140

I suppose I've already do that

lujiajing1126 commented:
> LGTM. Once test failures get fixed, we can merge this.

I've added a random prefix to the database dir and removed all of them after each test.

Besides, target `pkg` has been added to `make test`, which may explain why everything works well online.
